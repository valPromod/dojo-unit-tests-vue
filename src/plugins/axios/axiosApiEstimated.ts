import axios from "axios";

const instanceForApi = axios.create({
  baseURL: process.env.VUE_APP_API_BASE_URL_SHOP,
  headers: {
    "Accept-Language": "fr",
    "ocp-apim-subscription-key": process.env.VUE_APP_FRONT_ESTIMATED_MARKET_SUBSCRIPTION_KEY
  },
});
export default instanceForApi;
