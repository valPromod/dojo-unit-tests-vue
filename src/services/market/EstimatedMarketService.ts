import EstimatedShopsMarketBean from "@/model/market/EstimatedShopsMarketBean";
import PostEstimatedShopsMarketBean from "@/model/market/PostEstimatedShopsMarketBean";
import axiosApiEstimated from '@/plugins/axios/axiosApiEstimated';
import { AxiosPromise } from "axios";

export default class EstimatedMarketService {
    public static baseUrlEstimatedMarket = `/estimated-markets`;

    public static getEstimatedMarkets(line : string, markets : string, begin : string, end : string|undefined) : AxiosPromise<EstimatedShopsMarketBean[]> {
        return axiosApiEstimated.get(`${this.baseUrlEstimatedMarket}?line=${line}&markets=${markets}&begin=${begin}&end=${end? end : ''}`);
    }

    public static postEtimatedMarkets(body: PostEstimatedShopsMarketBean[]) :  AxiosPromise {
        return axiosApiEstimated.post(`${this.baseUrlEstimatedMarket}`, body);
    }
}
