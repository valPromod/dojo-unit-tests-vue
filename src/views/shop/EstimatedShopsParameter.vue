<template>
  <div>
    <div v-show="!isLoadingCriteria">
      <estimated-shops-criteria :marketsFromApi="segmentationsMarkets" :criteria="criteria" @search="search"/>
    </div>
    <v-divider></v-divider>
    <div class="alert">
      <v-alert
        v-for="(alert,index) in alerts"
        :key="index"
        :value="alert.show"
        :type="alert.type"
        colored-border
        dismissible
        elevation="2"
        class="ma-0"
      >
        {{ alert.message }}
      </v-alert>
    </div>

    <estimated-shops-data
      :estimatedShopsMarketOriginalList="datas"
      :yearWeeks="yearWeeks"
      :markets="criteria.markets"
      @save="save"
      v-if="datas && datas.length > 0 && !isLoadingData && !isLoadingCriteria"
    />

    <v-progress-circular
      :size="200"
      color="primary"
      indeterminate
      v-show="isLoadingData || isLoadingCriteria || isLoadingSave"
      class="loader"
    >Chargement
    </v-progress-circular>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import EstimatedShopsCriteria from "@/components/shop/EstimatedShopsParameterCriteria.vue";
import EstimatedShopsData from "@/components/shop/EstimatedShopsParameterData.vue";
import EstimatedShopsMarketBean from "@/model/market/EstimatedShopsMarketBean";
import SegmentationMarket from "@/model/shop/Segmentation";
import Criteria from "@/model/shop/EstimatedShopsParameterCriteria2";
import ShopSegmentationService from "@/services/shop/ShopSegmentation.service";
import EstimatedMarketService from "@/services/market/EstimatedMarketService";
import Alert from "@/model/alert/Alert2";
import PostEstimatedShopsMarketBean from "@/model/market/PostEstimatedShopsMarketBean";

@Component({
  components: {
    EstimatedShopsCriteria,
    EstimatedShopsData,
  },
})
export default class EstimatedShopsParameter extends Vue {

  public alerts : Alert[] = [];

  public isLoadingData = false;
  public isLoadingCriteria = false;
  public isLoadingSave = false;
  public segmentationsMarkets : SegmentationMarket[] = [];

  public criteria : Criteria = {};

  public yearWeeks : string[] = [];

  public datas? : EstimatedShopsMarketBean[] = [];
  
  created(): void {
    this.isLoadingCriteria = true;
    ShopSegmentationService.getShopSegmentation('market')
      .then(response => {
        this.segmentationsMarkets = response.data;
        if(this.segmentationsMarkets){
          let flatLinesFromMarket = this.segmentationsMarkets.map(market => market.line);
          let linesFromMarkets : string[] = flatLinesFromMarket.filter((line,pos) => flatLinesFromMarket.indexOf(line) == pos) as string[];
          this.criteria = {
            lines : linesFromMarkets.sort((a,b) => parseInt(a)-parseInt(b)),
            markets : this.segmentationsMarkets.filter(market => market.line === linesFromMarkets[0]),
          } 
        }
      })
      .catch(() => {
        this.alerts.push({show: true, message: "Erreur lors de la récupération des critères", type: "error"})
        })
      .finally(() => this.isLoadingCriteria = false);
  }

  search(line : string, markets : string[], begin : string, end : string|undefined) : void{
    this.isLoadingData = true;
    EstimatedMarketService.getEstimatedMarkets(line, markets.join(','), begin, end)
      .then(response => {
        this.datas = response.data;
        if(this.datas?.length){
          let market = this.datas[0];
          if(market?.rows.length){
            let size = market.rows[0];
            this.yearWeeks = size.yearWeeks.map(yearWeek => yearWeek.yearWeek);
          }
        }
      }).catch(() => {
        this.alerts.push({show: true, message: "Erreur lors de la récupération des données", type: "error"})
      }).finally(() => this.isLoadingData=false);
  }

  public getLineOfMarket(marketId : string) : string {
    if(this.criteria.markets){
      let market = this.criteria.markets.find(market => market.id === marketId);
      return market ? market.line : "";
    }
    return "";
  }

  public save(estimatedShopsMarketList : EstimatedShopsMarketBean[]): void {
    this.isLoadingSave = true;
    let body : PostEstimatedShopsMarketBean[] = estimatedShopsMarketList.map(estimatedMarket => {
        return {
            market: {
                id: estimatedMarket.market,
                line: this.getLineOfMarket(estimatedMarket.market)
            },
            rows: estimatedMarket.rows
        }
    })
    EstimatedMarketService.postEtimatedMarkets(body).then(() => {
      this.alerts.push({show: true, message: "Sauvegarde des données effectuée", type: "success"})
    })
      .catch(() => {
        this.alerts.push({show: true, message: "Erreur lors de la sauvegarde des données", type: "error"})
      }).finally(() => this.isLoadingSave=false);
  }
}

</script>
