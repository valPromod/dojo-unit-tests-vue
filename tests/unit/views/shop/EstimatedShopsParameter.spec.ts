import Vuetify from "vuetify";
import Vue from "vue";
import { createLocalVue, mount, shallowMount, Wrapper } from "@vue/test-utils";
import EstimatedShopsParameter from '@/views/shop/EstimatedShopsParameter.vue'
import ShopSegmentationService from "@/services/shop/ShopSegmentation.service";
import { mockAxiosPromise } from "../../../__utils__/axios";
import EstimatedMarketService from "@/services/market/EstimatedMarketService";

type RelaxedVue = Wrapper<
    EstimatedShopsParameter & { [key: string]: any }
>;


describe('EstimatedShopsParameter.vue', () => {

    const localVue = createLocalVue();
    let vuetify: Vuetify;

    const mountFunction = (options?: any, shallow = true): RelaxedVue => {
        const opts = {
            localVue,
            vuetify,
            ...options,
        };
        return shallow
            ? shallowMount(EstimatedShopsParameter, opts)
            : mount(EstimatedShopsParameter, opts);
    };

    beforeEach(() => {
        vuetify = new Vuetify();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });


    it('props pass well', () => {
        

        const spy = jest.spyOn(ShopSegmentationService, "getShopSegmentation").mockReturnValue(mockAxiosPromise([
            {
                id: "1",
                line: "1",
                wording: "wordingTest",
                segmentationType: "segmentationTest",
                order: 1,
                activite: "activiteTest"
            }]
        ));

        const spyGetEstimatedMarkets = jest.spyOn(EstimatedMarketService, "getEstimatedMarkets");

        const wrapper: RelaxedVue = mountFunction({}, false);

        wrapper.find('#id-test').trigger('click');
        expect(spy).toHaveBeenCalled();
        expect(spyGetEstimatedMarkets).not.toHaveBeenCalled();
    });

    /*it('props pass well', () => {
        

        const spy = jest.spyOn(ShopSegmentationService, "getShopSegmentation").mockReturnValue(mockAxiosPromise([
            {
                id: "1",
                line: "1",
                wording: "wordingTest",
                segmentationType: "segmentationTest",
                order: 1,
                activite: "activiteTest"
            }],
            500
        ));


        const wrapper: RelaxedVue = mountFunction();
        Vue.nextTick(() => {
            expect(wrapper.vm.alerts).toStrictEqual([{show: true, message: "Erreur lors de la récupération des critères", type: "error"}]);
        })
        
    });*/

    it("should not search without implantation value", async () => {
        const wrapper: RelaxedVue = mountFunction({
            propsData: {
            }
        }, false);
        Vue.nextTick(() => {
            expect(wrapper.html()).toMatchSnapshot();
        });
    });
})
