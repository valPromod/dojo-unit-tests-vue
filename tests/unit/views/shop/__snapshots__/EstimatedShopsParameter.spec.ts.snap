// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstimatedShopsParameter.vue should not search without implantation value 1`] = `
<div>
  <div style="display: none;">
    <div class="estimated-shops-parameter-criteria">
      <h1>Projection du parc Magasin</h1>
      <form novalidate="novalidate" class="v-form my-5">
        <div class="row">
          <div class="estimated-shops-parameter-criteria_column col">
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-60" class="v-input__slot">
                  <div class="v-select__slot"><label for="input-60" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Ligne</label>
                    <div class="v-select__selections">
                      <div class="v-select__selection v-select__selection--comma">1</div><input id="input-60" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                    </div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden" value="1">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-input theme--light v-text-field v-select v-select--chips v-select--is-multi">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-65" class="v-input__slot">
                  <div class="v-select__slot"><label for="input-65" class="v-label theme--light" style="left: 0px; position: absolute;">Marchés</label>
                    <div class="v-select__selections"><input id="input-65" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="estimated-shops-parameter-criteria_column col">
            <div class="d-flex align-center"><button type="button" class="v-icon notranslate v-icon--link mdi mdi-minus theme--light"></button>
              <div class="v-input mx-3 v-input--is-label-active v-input--is-dirty theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-71" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Début</label><input maxlength="6" required="required" id="input-71" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div> <button type="button" class="v-icon notranslate v-icon--link mdi mdi-plus theme--light"></button>
            </div>
            <div class="d-flex align-center"><button type="button" class="v-icon notranslate v-icon--link mdi mdi-minus theme--light"></button>
              <div class="v-input mx-3 theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-76" class="v-label theme--light" style="left: 0px; position: absolute;">Fin</label><input maxlength="6" id="input-76" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div> <button type="button" class="v-icon notranslate v-icon--link mdi mdi-plus theme--light"></button>
            </div>
          </div>
          <div class="spacer"></div> <button type="button" class="ma-5 v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round theme--light v-size--large primary" id="id-test"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light white--text">search</i></span></button>
        </div>
      </form>
    </div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="alert"></div>
  <!---->
  <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-circular loader v-progress-circular--visible v-progress-circular--indeterminate primary--text" style="height: 200px; width: 200px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="20.408163265306122 20.408163265306122 40.816326530612244 40.816326530612244" style="transform: rotate(0deg);">
      <circle fill="transparent" cx="40.816326530612244" cy="40.816326530612244" r="20" stroke-width="0.8163265306122449" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px" class="v-progress-circular__overlay"></circle>
    </svg>
    <div class="v-progress-circular__info">Chargement
    </div>
  </div>
</div>
`;
